[c10d] Add a new API for adding ephemeral timeout for one local rank and the timeout will reset when the first collective finishes

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #130905

We provide an API for user to add ephemeral timeout across all PGs within one rank and the timeout will reset when the first collective issued after the timeout added finishes.

Each extension only covers collectives after the issue and before the first collective finished. The diagram below shows how the timeout changes:

<img width="1174" alt="image" src="https://github.com/user-attachments/assets/354923b7-581c-40de-ae0f-1cd3da273ccc">


While this feature provides flexibility in specific scenarios, it introduces statefulness to timeout setting. Therefore, it is advisable to use this API sparingly and consider alternative approaches, such as directly setting the timeout or utilizing a barrier collective (one can set any timeout to the barrier), whenever feasible.
 



cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @wz337 @wconstab @d4l3k @c-p-i-o

fduwjj committed:
[c10d] Add a new API for timeout extend for one local rank

[ghstack-poisoned]

fduwjj commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -575,6 +575,33 @@ def test_set_nccl_pg_timeout(self, backend):
+    @requires_nccl()
+    @skip_but_pass_in_sandcastle_if(not TEST_MULTIGPU, "NCCL test requires 2+ GPUs")
+    @parametrize("backend", [None, "nccl"])
+    def test_extend_nccl_pg_timeout(self, backend):
```
This test somehow will fail periodically. So I have L593 which will give me different address on the python side.

fduwjj replied:
The reason test failed is that, after we set time, somehow when we get the backend and Options again, it now points to a different address and the timeout value was mess up. It is 701 secs instead of 600..

fduwjj replied:
Solved.

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroup.cpp:
```diff
@@ -89,6 +89,8 @@ c10::intrusive_ptr<Backend> ProcessGroup::getBackend(
     c10::DeviceType deviceType) {
   // If there is a backend associated with this device type then return it
   if (deviceTypeToBackend_.find(deviceType) != deviceTypeToBackend_.end()) {
+    LOG(ERROR) << "Address of return: " << &deviceTypeToBackend_[deviceType]
```
Somehow for the same rank and same pg, I do see different address here. Is this expected? @H-Huang @kwen2501

H-Huang replied:
Do you mean for the same device, if so that would be strange? Or is it just different addresses because you are passing different devices.

fduwjj replied:
OK, the finding is that, we indeed return the same address here but somehow the pybind makes it not same when calling id(var) in python, if we print(var) directly it shows same address as well.

H-Huang commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1297,6 +1298,65 @@ def get_node_local_rank(fallback_rank: Optional[int] = None) -> int:
+    assert isinstance(group, ProcessGroup)
+    devices = group._device_types
+    backends = set()
+    if torch.device("cpu") in devices and is_gloo_available():
```
kind of confused about all the if-statement checks. Is there a downside in simplifying all the logic it to just:

```
for device in devices:
    backends.add(group._get_backend(device))
```

fduwjj replied:
Oh because we only have pybind API defined for NCCL and Gloo backend, unfortunately. Or do we want to actually enable them for all backend?

H-Huang commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1297,6 +1298,65 @@ def get_node_local_rank(fallback_rank: Optional[int] = None) -> int:
+
+def _extend_timeout_all_pgs(timeout: timedelta) -> None:
+    """
+    This API extends the timeout for all PGs locally on one rank.
```
just curious, is the intended use case for just one rank to call this? Or do we expect all ranks to call this.

fduwjj replied:
Just for one or couple ranks not all ranks. At least this is the requirement from PT2 folks.

fduwjj committed:
Update on "[RFC and not for landing now][c10d] Add a new API for timeout extend for one local rank"


We provide an API for user to extend and reset timeout across all PGs within one rank.


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj committed:
Update on "[RFC and not for landing now][c10d] Add a new API for timeout extend for one local rank"


We provide an API for user to extend and reset timeout across all PGs within one rank.


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

shuqiangzhang commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1297,6 +1298,65 @@ def get_node_local_rank(fallback_rank: Optional[int] = None) -> int:
     )
 
 
+def _get_backends_for_timeout_set(group: ProcessGroup) -> Set[c10d_Backend]:
```
Curious, do we really have use cases where a single PG has multiple backends?

fduwjj replied:
when users does not pass in backend specifically, there are multiple backends. I think @H-Huang can comment more on this.

ezyang replied:
This seems kind of overwrought given that we don't actually support setting ephemeral timeout on anything that is not NCCL lol. I guess it is symmetric with the current timeout setting logic though.

fduwjj replied:
Yeah, we can just get backend for nccl here. So we don't touch the old python code. Do we care about other backends like gloo and MPI, probably not...

shuqiangzhang commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1297,6 +1298,65 @@ def get_node_local_rank(fallback_rank: Optional[int] = None) -> int:
+    """
+    This API extends the timeout for all PGs locally on one rank.
+    If user sets a very large negative timeout, we will set the timeout
+    to be the negative value.
```
Hmm, we should never allow a negative timeout?

wconstab replied:
I'm not sure it's a good idea to let people subtract from the timeout.

shuqiangzhang commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1297,6 +1298,65 @@ def get_node_local_rank(fallback_rank: Optional[int] = None) -> int:
+            if timeout + current_timeout > zero_delta:
+                backend._set_default_timeout(timeout + current_timeout)
+            else:
+                reset_timeout = (
```
I think in this case, we should warn users that his input value is not really valid and report errors, instead of setting the timeout to default that which silently goes against users' intention.

wconstab commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -575,6 +575,38 @@ def test_set_nccl_pg_timeout(self, backend):
+            self._check_nccl_timeout(timedelta(seconds=103))
+            pg.allreduce(torch.rand(10).cuda(self.rank))
+            dist.distributed_c10d._extend_timeout_all_pgs(timedelta(seconds=-10000))
+            self._check_nccl_timeout(timedelta(seconds=600))
```
I'm not sure it's very intuitive to do 103 - 10000 and get 600 as a result. What's going on here?

fduwjj replied:
Well, it's for users to reset the timeout and the default one for NCCL is 600, as mentioned in the RFC doc.

wconstab commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1297,6 +1298,65 @@ def get_node_local_rank(fallback_rank: Optional[int] = None) -> int:
+                backend._set_default_timeout(reset_timeout)
+
+
 def _set_pg_timeout(timeout: timedelta, group: Optional[ProcessGroup] = None) -> None:
```
I'm not really sure why we need to offer a new api. Users could directly call set_pg_timeout and modify it to a value they want. 

Can they also get the current timeout? Then they can do all the crazy math and bookkeeping.

fduwjj replied:
I guess the extra we provide here is let users to set timeout for all subPGs and do we want users to iterate through PGs themselves?

fduwjj committed:
Update

[ghstack-poisoned]

fduwjj committed:
Update

[ghstack-poisoned]

fduwjj committed:
Update

[ghstack-poisoned]

fduwjj commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -578,6 +578,36 @@ def test_set_nccl_pg_timeout(self, backend):
+            time.sleep(6)
+            pg.allreduce(torch.rand(10).cuda(self.rank))
+        else:
+            dist.distributed_c10d._extend_timeout_until_first_done_all_pgs(
```
currently we don't test the allreduce finished and no timeout.. Might need more tests on this.

ezyang commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -839,6 +846,17 @@ class TORCH_API ProcessGroupNCCL : public Backend {
+
+  // Record of first work and will get reset if timeout is further extended.
+  c10::intrusive_ptr<ProcessGroupNCCL::WorkNCCL> firstWorkSinceExtended_ =
+      nullptr;
```
My comment from the Phabricator diff still applies here https://www.internalfb.com/diff/D60274726?transaction_fbid=1141409296951808

> You don't need an entire queue. Let's assume that NCCL collectives complete in the order they were enqueued. You need only two quantities: "accumulated ephemeral timeout" (total_timeout) and "timeout pending reset" (inflight_timeout). Whenever you extend the timeout, you increase the accumulated timeout. Whenever you issue a collective, you record timeout to reset (work_timeout_reduce) after this collective completes as "total_timeout - inflight_timeout" on the Work object, and then set inflight_timeout := total_timeout. The invariant is that the sum of all work_timeout_reduce on outstanding Work objects is equal to inflight_timeout. When the work completes, you reduce total_timeout and inflight_timeout by the work_timeout_reduce. Eventually, you will quiesce back to zero total_timeout.

In particular, I do not think you should need to store the first work since extended. It seems more logical to associate the timeout pending reset ON the work object itself, this allows for multiple pending work with different resets.

fduwjj replied:
This proposal also works and I think I get it. But why would a Work object care about `work_timeout_reduce` it's essentially a concept outside Work? What each work wants to know is what is my own timeout, IMHO.

fduwjj replied:
Actually we draw this one to ensure we understand you correctly:

<img width="1173" alt="image" src="https://github.com/user-attachments/assets/e9e64688-edf8-497b-bb0d-4f31759e3fd1">


Yeah this indeed is better than the current one because if there are interleave of collective finishing and timeout extension, we don't bump up timeout monotonically. But instead of storing it into Work object, we want to use a map to store.

ezyang replied:
Yup, your table looks right to me.

If you want to argue about storing it on Work or storing it in a map, well, it's all semantically isomorphic in the end so I don't really care. I had been coming at it from the perspective that Work was a specifically c10d collective concept though. Note that if you do a map you have to worry about (1) lifetimes, (2) ABA problems, (3) it's a little less efficient because you have to do the lookup, (4) locking.

fduwjj replied:
Got it. I agree, we need to handle the lifetimes for the map and it has extra cost.. But we will need a lock anyway even if we store the reduce_time into work object and with lock, there shouldn't be an ABA problem because each time, read/write happens inside a mutex which makes it atomic?

ezyang replied:
I'm definitely not saying the other problems aren't solvable, just need to make sure the i's are dotted and the t's crossed

fduwjj committed:
Update on "[RFC][c10d] Add a new API for timeout extend for one local rank and the timeout will reset when the first collective finishes"


We provide an API for user to extend and reset timeout across all PGs within one rank and the timeout will reset when the first collective finishes.


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj committed:
Update on "[RFC][c10d] Add a new API for timeout extend for one local rank and the timeout will reset when the first collective finishes"


We provide an API for user to extend and reset timeout across all PGs within one rank and the timeout will reset when the first collective finishes.

Each extension only covers collectives after the issue and before the first collective finished. The diagram below shows how the timeout changes:

<img width="1174" alt="image" src="https://github.com/user-attachments/assets/354923b7-581c-40de-ae0f-1cd3da273ccc">
 



cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj committed:
Update on "[RFC][c10d] Add a new API for timeout extend for one local rank and the timeout will reset when the first collective finishes"


We provide an API for user to extend and reset timeout across all PGs within one rank and the timeout will reset when the first collective finishes.

Each extension only covers collectives after the issue and before the first collective finished. The diagram below shows how the timeout changes:

<img width="1174" alt="image" src="https://github.com/user-attachments/assets/354923b7-581c-40de-ae0f-1cd3da273ccc">
 



cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj committed:
Update on "[RFC][c10d] Add a new API for timeout extend for one local rank and the timeout will reset when the first collective finishes"


We provide an API for user to extend and reset timeout across all PGs within one rank and the timeout will reset when the first collective finishes.

Each extension only covers collectives after the issue and before the first collective finished. The diagram below shows how the timeout changes:

<img width="1174" alt="image" src="https://github.com/user-attachments/assets/354923b7-581c-40de-ae0f-1cd3da273ccc">
 



cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -2989,34 +3001,45 @@ such as `dist.all_reduce(tensor, async_op=True)`.
+            return ivalue.toCustomClass<::c10d::Work>();
+          });
+
+#ifdef USE_C10D_NCCL
+  py::class_<
+      ::c10d::ProcessGroupNCCL::WorkNCCL,
+      std::shared_ptr<::c10d::ProcessGroupNCCL::WorkNCCL>>(
+      processGroupNCCL,
+      "WorkNCCL");
+//       R"(
+// A `WorkNCCL` object represents the handle to a `Work` object issued on GPU.)")
+    //   .def("timeout", &::c10d::ProcessGroupNCCL::WorkNCCL::getOpTimeout);
+#endif
```
I need to expose WorkNCCL via pybind, but as long as I have this piece of code, it will just fail init of c10d...

wconstab replied:
why do you need to expose Work via pybind?

it makes me nervous because we already expose Work but maybe its not via pybind- maybe its via torchbind (the thing from torchscript interpreter)?

I ran into some gnarly issues with binding incompatibility when working on functional collectives.  @howard was helpful.  In any case, i'm hoping you can just avoid having to bind WorkNCCL?

fduwjj replied:
Well, I implicitly check the timeout change is expected, but we need a way to check the timeout of a Work? Maybe I can create an API inside NCCLPG?

howard replied:
> @howard was helpful

I think you mean someone else 😉

fduwjj replied:
cc: @H-Huang

ezyang commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -341,6 +344,9 @@ class TORCH_API ProcessGroupNCCL : public Backend {
     std::chrono::milliseconds opTimeout_;
 
+    // timeout to be reduced after this work finishes.
+    std::chrono::milliseconds opTimeoutReduce_ = std::chrono::milliseconds(0);
```
opTimeout totally fails to convey "timeout to be reduced after this work finishes"

fduwjj replied:
Do you have any suggestions on a better name? Will `timeoutReduceAfterFinish_` too long?

ezyang replied:
You can make up a new term of art that doesn't have bad connotations, like `ephemeralTimeoutInflight_`, a long name isn't bad either but if you've got some unified concept that shows up a lot of places, having a consistent keyword for it helps a bit.

ezyang commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -642,6 +648,8 @@ class TORCH_API ProcessGroupNCCL : public Backend {
 
   void performNocolorSplit(at::Device device);
 
```
nb: no longer modification to Work

fduwjj replied:
No, we have WorkNCCL and the timeout is actually added there...

ezyang commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -642,6 +648,8 @@ class TORCH_API ProcessGroupNCCL : public Backend {
 
   void performNocolorSplit(at::Device device);
 
+  void extendTimeoutUntilFirstDone(const std::chrono::milliseconds& timeout);
```
It's probably better to just come up with a term of art for it (like "ephemeral timeout") and then explain what it is in docs.

fduwjj replied:
If you call it "ephemeral", but what does it mean when we call this API multiple times which is very much likely?

ezyang replied:
The timeout is still ephemeral, in the sense that any timeout you add via this API eventually goes away (even if you call it multiple times).

ezyang commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -801,6 +809,11 @@ class TORCH_API ProcessGroupNCCL : public Backend {
+  // Util function to assign timeout to each work.
+  void assignTimeoutToWork(
+      c10::intrusive_ptr<ProcessGroupNCCL::WorkNCCL> work,
+      c10::intrusive_ptr<Options> option);
```
Options 🤔 

Are you sure you want to take ownership for these? Seems unlikely...

fduwjj replied:
They are passed in from ctor. We don't take ownership of Options.

ezyang replied:
If you don't take ownership, pass by const ref

fduwjj replied:
Sure.

ezyang commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -839,6 +852,17 @@ class TORCH_API ProcessGroupNCCL : public Backend {
 
+  // The lock which protects the write/read of
+  // extendedTimeout_/inflightTimeoutExt_.
+  std::mutex mtxTimeoutExtension_;
```
As you already have a mutex

```
  std::mutex mutex_;
```

You really need to explain why you need a second one, since two mutexes opens up the possibility for deadlock. It's not free!

fduwjj replied:
Honestly `std::mutex mutex_;` steals the generic name..  It is mostly used when it comes to exception handling and `devNCCLCommMap_`, etc...

OK, I think we might need to clean up and try to control number of mutex here. We can reuse that mutex_ instead of creating a new one.

fduwjj replied:
Wait, this `mutex_` is used when doing NCCLCommAbort which can have hang...We probably cannot use it. But I agree, we need to do an audit to the mutex here and see if we can consolidate them.

ezyang commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -839,6 +852,17 @@ class TORCH_API ProcessGroupNCCL : public Backend {
+  std::chrono::milliseconds extendedTimeout_ = std::chrono::milliseconds(0);
+
+  // The timeout extension which has been already applied to work.
+  std::chrono::milliseconds inflightTimeoutExt_ = std::chrono::milliseconds(0);
```
It's weird that the first variable is named extendedTimeout but the second is TimeoutExt. Pick... a consistent ordering?

fduwjj replied:
Sure. Originally, I want to just call it `inflightTimeout_` but here it only means the extended part.. OK. I can make it consistent.

shuqiangzhang replied:
We'd better minimize the states and so many different definition of timeouts make things hard to read. Is this additional timeout necessary? extendedTimeout would be set by user, and reset to 0 when the first work completes. When checkout timeout and extendedTimeout_ > 0,  you can pass the customized timeout to ProcessGroupNCCL::WorkNCCL::checkTimeout().

ezyang replied:
This one is needed, because you aren't supposed to necessarily reset to zero when the first work completes. The interleaved case can be seen in the spreadsheet in PR desc.

ezyang commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2409,6 +2428,19 @@ uint64_t ProcessGroupNCCL::WorkNCCL::getSequencenumber() const {
+    c10::intrusive_ptr<ProcessGroupNCCL::WorkNCCL> work,
+    c10::intrusive_ptr<ProcessGroupNCCL::Options> option) {
+  std::chrono::milliseconds timeout = option->timeout;
+  std::unique_lock<std::mutex> lock(mtxTimeoutExtension_);
```
What's locking discipline on a PG overall? If I mutably access it from multiple threads is it expected that I implicitly have taken out another lock, or is PG responsible for locking? Latter is pretty unusual for PyTorch C++ API design

fduwjj replied:
I don't think I am following the question here, are you asking whether the lock is implicitly acquired or explicitly lock/unlock? If you are asking about this, I am afraid PG need to be responsible for the lock...

ezyang replied:
So, let's say you design a struct in C++ with some methods, and this struct is potentially accessible from multiple threads. You have to decide what locking strategy is. One approach is to say mutations from multiple threads not allowed, you have to make sure you never race mutations. at::Tensor works this way; multithread access OK, but not multithread writes. Another approach is to modify every method to implicitly take out a lock when you call it, so method calls are thread safe without any special synchronization. This sort of thing is very popular in Java because they have a dedicated keyword synchronized that does it.

fduwjj replied:
I see. Thanks for the detailed explanation. I think in PGNCCL we unfortunately don't do single thread write (aka the at::Tensor way you mentioned here) for example, we let the main thread to add the Work into queue and then erase it in the watchdog thread. For this PR specifically, we also do the similar thing, for example in `ephemeralTimeoutSet_` we set it in the main thread but reset in the watchdog thread. I guess the reason here is because we use watchdog thread as a side thread to do all cleanups and completion check. And the mutex we add here is like what you mentioned here as a implicit lock. (watchdog only calls on function, lol)

fduwjj committed:
Update on "[RFC][c10d] Add a new API for timeout extend for one local rank and the timeout will reset when the first collective finishes"


We provide an API for user to extend and reset timeout across all PGs within one rank and the timeout will reset when the first collective finishes.

Each extension only covers collectives after the issue and before the first collective finished. The diagram below shows how the timeout changes:

<img width="1174" alt="image" src="https://github.com/user-attachments/assets/354923b7-581c-40de-ae0f-1cd3da273ccc">
 



cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1777,6 +1788,14 @@ void ProcessGroupNCCL::watchdogHandler() {
       if (work.isCompleted()) {
+        {
+          // Reset the timeout and first work if the work is completed.
+          std::unique_lock<std::mutex> lock(mtxTimeoutExtension_);
```
Opps, what I want to use is `std::lock_guard`. `std::lock_guard` comes with automatic unlock.

fduwjj committed:
Update on "[RFC][c10d] Add a new API for timeout extend for one local rank and the timeout will reset when the first collective finishes"


We provide an API for user to extend and reset timeout across all PGs within one rank and the timeout will reset when the first collective finishes.

Each extension only covers collectives after the issue and before the first collective finished. The diagram below shows how the timeout changes:

<img width="1174" alt="image" src="https://github.com/user-attachments/assets/354923b7-581c-40de-ae0f-1cd3da273ccc">
 



cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj committed:
Update on "[RFC][c10d] Add a new API for timeout extend for one local rank and the timeout will reset when the first collective finishes"


We provide an API for user to extend and reset timeout across all PGs within one rank and the timeout will reset when the first collective finishes.

Each extension only covers collectives after the issue and before the first collective finished. The diagram below shows how the timeout changes:

<img width="1174" alt="image" src="https://github.com/user-attachments/assets/354923b7-581c-40de-ae0f-1cd3da273ccc">
 



cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

wconstab commented on torch/_C/_distributed_c10d.pyi:
```diff
@@ -271,17 +271,29 @@ class Work:
     def unbox(obj: ScriptObject) -> Work: ...
 
 class Backend:
+    class Options:
```
why did you have to move Options into Backend? I don't see why options is relevant to any of the other apis

fduwjj replied:
No, it's because Backend has a option. The options in NCCLPG inherit from that:

Code is here:
https://github.com/pytorch/pytorch/blob/596f568592ad312fb85f22cca1618af8d327f859/torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp#L413

shuqiangzhang commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1777,6 +1796,14 @@ void ProcessGroupNCCL::watchdogHandler() {
+          // Reset the timeout and first work if the work is completed.
+          std::lock_guard<std::mutex> lock(mtxTimeoutExtension_);
+          if (work.opTimeoutReduce_.count() > 0) {
+            extendedTimeout_ -= work.opTimeoutReduce_;
```
Do we really need to keep timeout states in Work objects, opTimeoutReduce_? ProcessGroupNCCL::WorkNCCL::checkTimeout() already allows you to pass an timeout value to replace the default one

fduwjj replied:
Instead of using a map to store of Work and timeoutReductionOnComplete_, storing in WorkNCCL is pretty much similar.

You can check the discussion here:
https://github.com/pytorch/pytorch/pull/130905#discussion_r1695416936

fduwjj committed:
Update on "[c10d] Add a new API for timeout extend for one local rank and the timeout will reset when the first collective finishes"


We provide an API for user to extend and reset timeout across all PGs within one rank and the timeout will reset when the first collective finishes.

Each extension only covers collectives after the issue and before the first collective finished. The diagram below shows how the timeout changes:

<img width="1174" alt="image" src="https://github.com/user-attachments/assets/354923b7-581c-40de-ae0f-1cd3da273ccc">
 



cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

ezyang commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1602,6 +1603,24 @@ const std::vector<uint64_t>& ProcessGroupNCCL::groupRanks() const {
   return options_->global_ranks_in_group;
 }
 
+void ProcessGroupNCCL::setEphemeralTimeout(
```
Wouldn't add be better than set, as you are accumulating ephemeral timeout here?

ezyang commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1602,6 +1603,24 @@ const std::vector<uint64_t>& ProcessGroupNCCL::groupRanks() const {
+void ProcessGroupNCCL::setEphemeralTimeout(
+    const std::chrono::milliseconds& timeout) {
+  std::lock_guard<std::mutex> lock(mtxTimeoutExtension_);
+  ephemeralTimeoutSet_ += timeout;
```
Maybe not set? Sounds like a hash set. `ephemeralTimeoutActive_`?

ezyang commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1602,6 +1603,24 @@ const std::vector<uint64_t>& ProcessGroupNCCL::groupRanks() const {
+  // Since collective returns a c10d::Work, we need to cast it to WorkNCCL.
+  if (auto workNCCL = c10::dynamic_intrusive_pointer_cast<WorkNCCL>(work)) {
+    // workNCCL is now a c10::intrusive_ptr<WorkNCCL>
+    return workNCCL->opTimeout_ == timeout;
```
Instead of doing this comparison, why don't you just make an accessor for `opTimeout_`? Or if you're going to keep this as is, it would be good to give it a name that clearly indicates that this is for debugging/testing purposes only. "check" implies that there's some sort of error checking going on in the inside, which there isn't.

fduwjj replied:
> why don't you just make an accessor for opTimeout_?

Great question, we don't expose `WorkNCCL` via pybind. And I tried to do it but it lead to distribute initialization failure. Sure let me give it a more accurate name.

ezyang commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -2706,6 +2707,24 @@ options :class:`~torch.distributed.ProcessGroupNCCL.Options`).
+              },
+              py::arg("work"),
+              py::arg("timeout"),
+              py::call_guard<py::gil_scoped_release>())
```
just fyi: these are very cheap functions, if you weren't taking locks inside them it wouldn't be necessary to release GIL on them

ezyang commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1777,6 +1796,14 @@ void ProcessGroupNCCL::watchdogHandler() {
       if (work.isCompleted()) {
+        {
+          // Reset the timeout and first work if the work is completed.
+          std::lock_guard<std::mutex> lock(mtxTimeoutExtension_);
```
NB, you have  `std::unique_lock<std::mutex> lock(workMetaListMutex_);` lock taken out right now, potential deadlock here

ezyang replied:
It looks like we never take out a lock while holding this lock, so this should be sufficient

fduwjj replied:
I thought the compiler will complain about it... lol let's use a different name here.

ezyang reviewed (approved):
This looks like it will work.

fduwjj committed:
Update on "[c10d] Add a new API for timeout extend for one local rank and the timeout will reset when the first collective finishes"


We provide an API for user to extend and reset timeout across all PGs within one rank and the timeout will reset when the first collective finishes.

Each extension only covers collectives after the issue and before the first collective finished. The diagram below shows how the timeout changes:

<img width="1174" alt="image" src="https://github.com/user-attachments/assets/354923b7-581c-40de-ae0f-1cd3da273ccc">
 



cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1602,6 +1603,24 @@ const std::vector<uint64_t>& ProcessGroupNCCL::groupRanks() const {
   return options_->global_ranks_in_group;
 }
 
+void ProcessGroupNCCL::addEphemeralTimeout(
```
nit: update PR desc to mention ephemeral and include more notes about how it works and what its for

fduwjj replied:
Sure, I added more details to PR desc.

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2409,6 +2436,20 @@ uint64_t ProcessGroupNCCL::WorkNCCL::getSequencenumber() const {
+    timeout += ephemeralTimeoutActive_;
+  }
+  work->opTimeout_ = timeout;
+  work->timeoutReductionOnComplete_ =
```
nit: rename to work->ownedEphermeralTimeout ?

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -341,6 +341,10 @@ class TORCH_API ProcessGroupNCCL : public Backend {
     // Clone of opTimeout_ from ProcessGroupNCCL.
     std::chrono::milliseconds opTimeout_;
 
+    // timeout to be reduced after this work finishes.
```
more verbose explanation

"ephemeral timeouts are owned by exactly one work, and decay after that work completes.  There may be more than one ephemeral timeout active at the same time." 

or something like hat

fduwjj committed:
Update on "[c10d] Add a new API for timeout extend for one local rank and the timeout will reset when the first collective finishes"


We provide an API for user to add ephemeral timeout across all PGs within one rank and the timeout will reset when the first collective issued after the timeout added finishes.

Each extension only covers collectives after the issue and before the first collective finished. The diagram below shows how the timeout changes:

<img width="1174" alt="image" src="https://github.com/user-attachments/assets/354923b7-581c-40de-ae0f-1cd3da273ccc">


While this feature provides flexibility in specific scenarios, it introduces statefulness to timeout setting. Therefore, it is advisable to use this API sparingly and consider alternative approaches, such as directly setting the timeout or utilizing a barrier collective (one can set any timeout to the barrier), whenever feasible.
 



cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Add a new API for timeout extend for one local rank and the timeout will reset when the first collective finishes"


We provide an API for user to add ephemeral timeout across all PGs within one rank and the timeout will reset when the first collective issued after the timeout added finishes.

Each extension only covers collectives after the issue and before the first collective finished. The diagram below shows how the timeout changes:

<img width="1174" alt="image" src="https://github.com/user-attachments/assets/354923b7-581c-40de-ae0f-1cd3da273ccc">


While this feature provides flexibility in specific scenarios, it introduces statefulness to timeout setting. Therefore, it is advisable to use this API sparingly and consider alternative approaches, such as directly setting the timeout or utilizing a barrier collective (one can set any timeout to the barrier), whenever feasible.
 



cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Add a new API for adding ephemeral timeout for one local rank and the timeout will reset when the first collective finishes"


We provide an API for user to add ephemeral timeout across all PGs within one rank and the timeout will reset when the first collective issued after the timeout added finishes.

Each extension only covers collectives after the issue and before the first collective finished. The diagram below shows how the timeout changes:

<img width="1174" alt="image" src="https://github.com/user-attachments/assets/354923b7-581c-40de-ae0f-1cd3da273ccc">


While this feature provides flexibility in specific scenarios, it introduces statefulness to timeout setting. Therefore, it is advisable to use this API sparingly and consider alternative approaches, such as directly setting the timeout or utilizing a barrier collective (one can set any timeout to the barrier), whenever feasible.
 



cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge
